File: frontend/src/api/blog.ts
import api from "./client";
import adminApi from "./admin";

// ── Public types ──────────────────────────────────────────────────────────────

export interface BlogArticleCard {
  id: number;
  title: string;
  slug: string;
  excerpt: string | null;
  cover_image: string | null;
  category: string;
  tags: string[];
  author: string;
  published_at: string | null;
  view_count: number;
  like_count: number;
}

export interface BlogArticleDetail extends BlogArticleCard {
  content: string | null;
  meta_title: string | null;
  meta_description: string | null;
  status: string;
  created_at: string | null;
  updated_at: string | null;
}

export interface BlogListResponse {
  items: BlogArticleCard[];
  total: number;
  page: number;
  pages: number;
}

// ── Public API ────────────────────────────────────────────────────────────────

export const fetchBlogArticles = (params: { page?: number; category?: string; search?: string }) =>
  api.get<BlogListResponse>("/blog/articles", { params }).then((r) => r.data);

export const fetchBlogArticle = (slug: string) =>
  api.get<BlogArticleDetail>(`/blog/articles/${slug}`).then((r) => r.data);

export const registerView = (id: number) =>
  api.post<{ view_count: number }>(`/blog/articles/${id}/view`).then((r) => r.data);

export const toggleLike = (id: number) =>
  api.post<{ like_count: number; liked: boolean }>(`/blog/articles/${id}/like`).then((r) => r.data);

// ── Admin types ───────────────────────────────────────────────────────────────

export interface ArticleCreatePayload {
  title: string;
  slug: string;
  excerpt?: string;
  content?: string;
  cover_image?: string;
  category: string;
  tags: string[];
  author: string;
  published_at?: string | null;
  status: string;
  meta_title?: string;
  meta_description?: string;
}

export type ArticleUpdatePayload = Partial<ArticleCreatePayload>;

// ── Admin API ─────────────────────────────────────────────────────────────────

export const adminFetchArticles = (params?: { status?: string; category?: string }) =>
  adminApi.get<BlogArticleDetail[]>("/blog/articles", { params }).then((r) => r.data);

export const adminFetchArticle = (id: number) =>
  adminApi.get<BlogArticleDetail>(`/blog/articles/${id}`).then((r) => r.data);

export const adminCreateArticle = (payload: ArticleCreatePayload) =>
  adminApi.post<BlogArticleDetail>("/blog/articles", payload).then((r) => r.data);

export const adminUpdateArticle = (id: number, payload: ArticleUpdatePayload) =>
  adminApi.put<BlogArticleDetail>(`/blog/articles/${id}`, payload).then((r) => r.data);

export const adminDeleteArticle = (id: number) =>
  adminApi.delete(`/blog/articles/${id}`);

export const adminUploadMedia = (file: File) => {
  const form = new FormData();
  form.append("file", file);
  return adminApi
    .post<{ url: string }>("/blog/media/upload", form, {
      headers: { "Content-Type": "multipart/form-data" },
    })
    .then((r) => r.data.url);
};

File: services/api-gateway/app/schemas/blog.py
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, field_validator


class ArticleCardOut(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    title: str
    slug: str
    excerpt: Optional[str] = None
    cover_image: Optional[str] = None
    category: str
    tags: list[str] = []
    author: str
    published_at: Optional[datetime] = None
    view_count: int = 0
    like_count: int = 0


class ArticleDetailOut(ArticleCardOut):
    content: Optional[str] = None
    meta_title: Optional[str] = None
    meta_description: Optional[str] = None
    status: str = "draft"
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class ArticleListResponse(BaseModel):
    items: list[ArticleCardOut]
    total: int
    page: int
    pages: int


class ArticleCreateRequest(BaseModel):
    title: str
    slug: str
    excerpt: Optional[str] = None
    content: Optional[str] = None
    cover_image: Optional[str] = None
    category: str = "general"
    tags: list[str] = []
    author: str = "Команда Attribly"
    published_at: Optional[datetime] = None
    status: str = "draft"
    meta_title: Optional[str] = None
    meta_description: Optional[str] = None


class ArticleUpdateRequest(BaseModel):
    title: Optional[str] = None
    slug: Optional[str] = None
    excerpt: Optional[str] = None
    content: Optional[str] = None
    cover_image: Optional[str] = None
    category: Optional[str] = None
    tags: Optional[list[str]] = None
    author: Optional[str] = None
    published_at: Optional[datetime] = None
    status: Optional[str] = None
    meta_title: Optional[str] = None
    meta_description: Optional[str] = None


class ViewResponse(BaseModel):
    view_count: int


class LikeResponse(BaseModel):
    like_count: int
    liked: bool


class MediaUploadResponse(BaseModel):
    url: str
